R/data-set.R

```diff
@@ -0,0 +1,141 @@
+      }
+
+      private$setColumnValues(private$.xValues, values)
+    },
+    #' @field yDimension Dimension in which the xValues are defined
+    yDimension = function(value) {
+      if (missing(value)) {
+        return(private$.yValues$dimension)
+      }
+      private$setColumnDimension(private$.yValues, value)
+    },
+    #' @field yUnit Unit in which the yValues are defined
+    yUnit = function(value) {
+      if (missing(value)) {
+        return(private$.yValues$displayUnit)
+      }
+      private$.yValues$displayUnit <- value
+    },
+    #' @field yValues Values stored in the yUnit dimension (not necessarily in the base unit of the dimension)
+    yValues = function(values) {
+      if (missing(values)) {
+        return(private$getColumnValues(private$.yValues))
+      }
+
+      private$setColumnValues(private$.yValues, values)
+    }
+  ),
+  public = list(
+    dataRepository = NULL,
+    #' @description
+    #' Initialize a new instance of the class
+    #' @param dataRepository Instance of the \code{DataRepository} object to wrap
+    #' @return A new `DotNetWrapper` object.
+    initialize = function(dataRepository = NULL) {
+      self$dataRepository <- dataRepository %||% private$createDataRepository()
+      private$initializeCache()
+    },
+    #' @description
+    #' Print the object to the console
+    #' @param ... Rest arguments.
+    print = function(...) {
+      private$printClass()
+      invisible(self)
+    }
+  ),
+  private = list(
+    .xValues = NULL,
+    .yValues = NULL,
+    setColumnValues = function(column, values) {
+      # values are set in the display unit. We need to make sure we convert them to the base unit
+      valuesInBaseUnit <- toBaseUnit(quantityOrDimension = column$dimension, values = values, unit = column$displayUnit)
+      column$values <- valuesInBaseUnit
+      invisible()
+    },
+    getColumnValues = function(column) {
+      # we need to convert the values in the display unit
+      toUnit(quantityOrDimension = column$dimension, values = column$values, targetUnit = column$displayUnit)
+    },
+    setColumnDimension = function(column, value){
+      # no need to update anything if we are setting the same values
+      if(column$dimension == value){
+        return()
+      }
+
+      # save the values in their display unit before updating
+      values <- private$getColumnValues(column)
+
+      #now we need to update dimension (display unit will be the default one as per .NET implementation)
+      column$dimension <- value
+
+      private$setColumnValues(column, values)
+    },
+    setColumnUnit = function(column, unit){
+      # no need to update anything if we are setting the same values
+      if(column$displayUnit == unit){
+        return()
+      }
+
+      # save the values in their display unit before updating
+      values <- private$getColumnValues(column)
+
+      #now we need to update dimension and display unit
+      column$displayUnit <- unit

```

Comment:
HERE: If the unit is not supported, we could easily update the dimension by using the method getDimensionForUnit. That way, the user would never have to worry about updating the dimension when updating the unit.

Reply:
@msevestre I would not prefer to automatically change the dimension, as I see it as a sanity check here. Simply changing the unit only affects what actual values the data have, but not the meaning of the data. Changing the dimension actually means that the data means something else. What do you think?